data_safe_haven/pulumi/dynamic/ssl_certificate.py

```diff
@@ -115,14 +114,17 @@ def create(self, props: dict[str, Any]) -> CreateResult:
             )
             outs["secret_id"] = kvcert.secret_id
         except Exception as exc:
-            msg = f"Failed to create SSL certificate [green]{props['certificate_secret_name']}[/] for [green]{props['domain_name']}[/].\n{exc!s}"
+            msg = (
```

Comment:
This might be clearer as `" ".join([...])` rather than needing to remember to leave whitespace in the strings.

Reply:
I disagree. If we use `" ".join` we can _only_ split on spaces.

I think string literal concatenation is idiomatic and efficient (concatenated at compile time rather than run time).

Reply:
Fair enough. There are a couple of places where you lost whitespace while (manually?) doing the splitting. Can you check?